### PR TITLE
[WIP] Replace Tpetra::CrsGraph with Tpetra::FECrsGraph

### PR DESCRIPTION
**Pull request by kinnewig**
tl;dr: It should be considered to replace the `Tpetra::CrsGraph`, `Tpetra::CrsMatrix`, and `Tpetra::Vector` in the new Tpetra-based  Trilinos interface of deal.II with `Tpetra::FECrsGraph`, `Tpetra::FECrsMatrix`, and `Tpetra::FEMultiVector` as it would bring a performance increase. But I don't have the time to do that at the moment.

Last week, @jpthiele and I participated in the European Trilinos User Group Meeting ([EuroTUG](https://eurotug.github.io/)). One session was about transitioning from Epetra to Tpetra. In that session, I learned about the classes `Tpetra::FECrsGraph`, `Tpetra::FECrsMatrix`, and `Tpetra::FEMultiVector`.

These classes are designed with finite element methods in mind. The `Tpetra::FE<>` uses the same idea that is already implemented in the deal.II in the `TpetraWrappers::Vector` class. 
All locally relevant elements can be accessed in the writable state as the IndexSet locally_relevant_dofs is used as the underlying index set in that state (the deal.II IndexSet corresponds to the Tpetra::Map). When `compress()`is called, the values in the ghost entries are communicated between the ranks, and the IndexSet locally_owned_index_set is employed as the underlying index set.

In the Tpetra interface of deal.II, we use that mechanism only for the vector and not for the sparse matrix at the moment. By using the `Tpetra::FE<>`, we can also bring that feature to the sparse matrix, which should result in a speed-up. 

As I'm currently busy writing my dissertation, I don't have time to replace the `Tpetra::CrsGraph`, `Tpetra::CrsMatrix`, and `Tpetra::MultiVector` with the corresponding `Tpetra::FE<>` classes, as this is quite some work. Once I have handed in my dissertation, I may have also had a short vacation, I will start working on this. From my point of view, this is not urgent at all, as the current implementation is working fine, and this will only increase performance. The only problem when using the Tpetra::FE classes is that we need more information, e.g.., one requires the locally_relevant_dofs IndexSet in the constructor and the reinit() functions.

### CODE REVIEW

**Comment by jpthiele** [comment]
And it should be noted that our initial choice and confusion came from a few things that were then unclear

* The Epetra::FECrsGraph used a linked list to store off-processor entries which was highly inefficient 
so that it was handled explicitly by deal.II
* Nevertheless, the Epetra::FE classes provided some more functionality that was needed, 
but was directly incorporated in the non FE classes of Epetra

Those two together with a very short and unclear documentation led us to believe that the FE classes were only there to make some search & replace things easier, when switching from Epetra to Tpetra.

**Comment by bangerth** [comment]
@kinnewig @jpthiele @gassmoeller In view of the upcoming release, how do we want to proceed here?

**Comment by gassmoeller** [comment]
I would say lets leave it open, but postpone to the next release. Replacing the classes would still be a good thing to try, in particular it could simplify our code quite a bit. But we already have the necessary functionality implemented ourselves, and replacing it would require a lot if work and performance testing.

**Comment by masterleinad** [comment]
Talking with Trilinos folks, they were saying that we wouldn't necessary benefit much but it's probably still worth a try.

**Comment by kinnewig** [comment]
> Talking with Trilinos folks, they were saying that we wouldn't necessary benefit much but it's probably still worth a try.

I was also told that switching to the `Tpetra::FE` probably brings no benefit, so I stopped working on this. Moreover, as I am no longer in academia, I am lacking the time to work on this.

**Comment by bangerth** [comment]
OK, then let's keep it open and milestone'd for the next release. @gassmoeller: If you felt compelled to benchmark this with your ASPECT testcases, perhaps that could resolve whether this is worth it.
